types/lodash/common/object.d.ts

```diff
@@ -2298,50 +2274,93 @@ declare module "../index" {
         * @param [thisArg] The `this` binding of `iteratee`.
         * @return Returns the new mapped object.
         */
+        mapValues<TResult>(obj: string | null | undefined, callback: StringIterator<TResult>): NumericDictionary<TResult>;
+
+        /**
+         * @see _.mapValues
+         * TODO: This would be better if we had a separate overload for obj: NumericDictionary that returned a NumericDictionary,
+         *       but TypeScript cannot select overload signatures based on number vs string index key type.
+         */
+        mapValues<T, TResult>(obj: Dictionary<T> | NumericDictionary<T> | null | undefined, callback: ObjectIterator<Dictionary<T>, TResult>): Dictionary<TResult>;
+
+        /**
+         * @see _.mapValues
+         */
         mapValues<T extends object, TResult>(obj: T | null | undefined, callback: ObjectIterator<T, TResult>): { [P in keyof T]: TResult };
 
+        /**
+         * @see _.mapValues
+         * TODO: This would be better if we had a separate overload for obj: NumericDictionary that returned a NumericDictionary,
+         *       but TypeScript cannot select overload signatures based on number vs string index key type.
+         */
+        mapValues<T>(obj: Dictionary<T> | NumericDictionary<T> | null | undefined, iteratee: object): Dictionary<boolean>;
+
         /**
          * @see _.mapValues
          */
         mapValues<T extends object>(obj: T | null | undefined, iteratee: object): { [P in keyof T]: boolean };
 
         /**
          * @see _.mapValues
+         * TODO: This would be better if we had a separate overload for obj: NumericDictionary that returned a NumericDictionary,
+         *       but TypeScript cannot select overload signatures based on number vs string index key type.
          */
-        mapValues<T, TKey extends keyof T>(obj: Dictionary<T> | null | undefined, iteratee: TKey): Dictionary<T[TKey]>;
+        mapValues<T, TKey extends keyof T>(obj: Dictionary<T> | NumericDictionary<T> | null | undefined, iteratee: TKey): Dictionary<T[TKey]>;
 
         /**
          * @see _.mapValues
+         * TODO: This would be better if we had a separate overload for obj: NumericDictionary that returned a NumericDictionary,
+         *       but TypeScript cannot select overload signatures based on number vs string index key type.
          */
-        mapValues<T extends object, TKey extends keyof T>(obj: T | null | undefined, iteratee: string): { [P in keyof T]: any };
+        mapValues<T>(obj: Dictionary<T> | NumericDictionary<T> | null | undefined, iteratee: string): Dictionary<T[keyof T]>;
```

Comment:
I think this should return `Dictionary<any>`. If we reach this overload, it's because `iteratee` does not match `keyof T`. So either they're specifying a property that doesn't exist, in which case the value will be `undefined`, or they're accessing some property that TypeScript doesn't know about, so we have no idea what the type will be.

Don't forget to update the wrapper versions, too.

Reply:
I was going to disagree with you for reasons that are not worth fully explaining, because I gathered from reading documentation that a `string` iteratee is not necessarily just a property name, but could also be a property "path" deeper into nested properties of the object. This support of "property paths" means we can't make any assumptions at all about the result type.